### PR TITLE
Further improvements to histogram

### PR DESCRIPTION
**Pull request by kdnilsen**
1. Replace "*" with "1" in each histogram column.  With this change, each histogram column reads as a binary number.

2. Adjust the span of histogram columns so that it ranges from the first non-empty tally of responses instead of sometimes starting with a negative response time.  With this improvement, the histogram provides higher resolution because each bucket spans less range.

3. Label each non-empty column of the histogram with the upper bound on response times for values corresponding to that column.

*Issue #, if available:*

*Description of changes:*
Improve histogram report

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by earthling-amzn** [comment]
I have a couple of style nits and a general comment. It looks like tabs and spaces might be mixed in this file? I'd also suggest always using braces \{,\} - even for one liners. I think it helps readability and can ease future maintenance work.

I've never seen histograms rendered like this before. I understand the rendering, but it's a bit harder for my brain to parse than what I'm used to seeing. What do you think about a more conventional ascii rendering, such as what Chrome uses (chrome://histograms/)? Chrome's histograms also incorporate percentiles. For example:

```
Histogram: CompositorLatency.DroppedFrame.CompositorAnimation.BeginImplFrameToSendBeginMainFrame recorded 8808 samples, mean = 3157.2 (flags = 0x41)
0      ... 
3      O                                                                         (1 = 0.0%) {0.0%}
4      -O                                                                        (3 = 0.0%) {0.0%}
5      --O                                                                       (6 = 0.1%) {0.0%}
6      -O                                                                        (5 = 0.1%) {0.1%}
8      -----O                                                                    (34 = 0.4%) {0.2%}
10     ----------O                                                               (94 = 1.1%) {0.6%}
13     -------------------------------------O                                    (474 = 5.4%) {1.6%}
17     ---------------------------------------------------O                      (823 = 9.3%) {7.0%}
22     -----------------------------------------------------------O              (951 = 10.8%) {16.3%}
28     -----------------------------------O                                      (565 = 6.4%) {27.1%}
36     -----------------------------------------------------O                    (858 = 9.7%) {33.6%}
46     -----------------------O                                                  (365 = 4.1%) {43.3%}
59     ---------O                                                                (143 = 1.6%) {47.4%}
76     ------O                                                                   (98 = 1.1%) {49.1%}
98     ------O                                                                   (89 = 1.0%) {50.2%}
127    ------------O                                                             (198 = 2.2%) {51.2%}
164    ------------------------------------O                                     (583 = 6.6%) {53.4%}
212    ------------------------------------------------------------------------O (1158 = 13.1%) {60.1%}
274    --------------------------------------------O                             (700 = 7.9%) {73.2%}
354    ------------------O                                                       (290 = 3.3%) {81.2%}
457    ---O                                                                      (54 = 0.6%) {84.4%}
590    --O                                                                       (38 = 0.4%) {85.1%}
762    --O                                                                       (38 = 0.4%) {85.5%}
984    ---O                                                                      (45 = 0.5%) {85.9%}
1270   ---O                                                                      (46 = 0.5%) {86.4%}
1640   -O                                                                        (21 = 0.2%) {87.0%}
2117   --O                                                                       (31 = 0.4%) {87.2%}
2733   --O                                                                       (28 = 0.3%) {87.5%}
3528   --O                                                                       (36 = 0.4%) {87.9%}
4555   ---O                                                                      (52 = 0.6%) {88.3%}
5880   ---O                                                                      (42 = 0.5%) {88.9%}
7591   --O                                                                       (33 = 0.4%) {89.3%}
9800   ---O                                                                      (48 = 0.5%) {89.7%}
12652  ----O                                                                     (60 = 0.7%) {90.3%}
16333  --O                                                                       (40 = 0.5%) {90.9%}
21085  ----O                                                                     (64 = 0.7%) {91.4%}
27220  -------------------------------------------O                              (688 = 7.8%) {92.1%}
35140  O                                                                         (5 = 0.1%) {99.9%}
45365  ... 
97606  O                                                                         (1 = 0.0%) {100.0%}
126007 ... 
```

**Comment by kdnilsen** [comment]
Thanks for feedback.  I should probably detabify all of those source files.  I had some bad editor options when I was creating much of this source.

We can change the form of histogram.  I actually am struggling to understand the Chrome format.  I gather that the line lengths are in some way logarithmic, but the scale is difficult for me to figure out.  (e.g. 94 looks like its about 3 times as long as 34, which would be appropriate for a linear scale, but 474 is only 30% longer than 94...). ???

**Comment by earthling-amzn** [comment]
I could accept some loss of precision in the scale as things are compressed to fit the column width, but the wildly different lengths just seems like a bug. Now that you've pointed it out, I see even worse discrepancies:
```
Histogram: Ads.ResourceUsage.Size.Cache2.Mainframe.AdResource recorded 225 samples, mean = 47.5 (flags = 0x41)
0    ------------------------------------------------------------------------O (57 = 25.3%)
1    -------------------O                                                      (15 = 6.7%) {25.3%}
2    -----O                                                                    (4 = 1.8%) {32.0%}
3    -O                                                                        (1 = 0.4%) {33.8%}
4    -O                                                                        (1 = 0.4%) {34.2%}
5    -----O                                                                    (4 = 1.8%) {34.7%}
6    --------------------O                                                     (32 = 14.2%) {36.4%}
8    --O                                                                       (3 = 1.3%) {50.7%}
10   -O                                                                        (3 = 1.3%) {52.0%}
13   O                                                                         (0 = 0.0%) {53.3%}
17   -------O                                                                  (27 = 12.0%) {53.3%}
22   --O                                                                       (7 = 3.1%) {65.3%}
29   -O                                                                        (5 = 2.2%) {68.4%}
38   ... 
64   -O                                                                        (2 = 0.9%) {70.7%}
83   -O                                                                        (5 = 2.2%) {71.6%}
108  -O                                                                        (5 = 2.2%) {73.8%}
141  -------------O                                                            (53 = 23.6%) {76.0%}
184  O                                                                         (1 = 0.4%) {99.6%}
```
53 is a depicted as less than 37 and much, much smaller than the first 57. So, let's just say that I find the chrome histograms easier to read, but something's clearly wrong with them.

**Comment by earthling-amzn** [comment]
After taking a closer look, I think the _height_ of the bar in the chrome histogram is actually the height of a rectangle that would cover the area defined by the bucket width and the sample count. In other words, wider buckets get shorter bars. This is not obvious, but I don't think it's a bug.

**Comment by kdnilsen** [comment]
Saying that the Chrome histogram is easier to read and at the same time admitting that you've never really understood what it represents is a bit "incongruent".  I'm still struggling to understand them.  I gather that you have made sense of it with your rectangle explanation, but don't yet "see it".  What is meant by the "height" of a horizontal bar.  Is height the row number?  And what are the units?

In any case, we can revisit at some future time the preferred representation of the histogram.  Would be nice to get a code review for the current implementation if someone can find time to do that so it can be upstreamed.

File: Extremem/src/main/java/com/amazon/corretto/benchmark/extremem/RelativeTimeMetrics.java [hunk 1]
```diff
@@ -1181,9 +1263,10 @@ static String us2s (ExtrememThread t, long us) {
     String result;
     long quotient = 0;
 
-    if (us == 0)
-      return "0us";
-    else if (us < 0) {
+    if (us == 0) {
+      Util.ephemeralString(t, 3);
+      return new String("0us");
```
Comment (earthling-amzn) [inline Extremem/src/main/java/com/amazon/corretto/benchmark/extremem/RelativeTimeMetrics.java:1268]: Why the explicit `new String` here?

**Review by AzeemJiva — APPROVED** [review]

**Pull request merged**